### ISSUE TITLE
Discord authentication cookie is not set only on Firefox, in Blazor server production

### ISSUE THREAD

**Issue by NexganGH** [aspnet-contrib/AspNet.Security.OAuth.Providers#576]
**Describe the bug**
Using blazor server, after trying to log in on the Discord page you get redirected to the correct route of the website but the correlation cookie is not set, getting these errors in console:
```'.AspNetCore.Correlation.wNmPU7i-6bJe81N8aurW-IvEZhT8aJjp1FHuy94h5hY' cookie not found.

Error from RemoteAuthentication: Correlation failed..

An unhandled exception has occurred while executing the request.

System.Exception: An error was encountered while handling the remote login.

 ---> System.Exception: Correlation failed.

   --- End of inner exception stack trace ---

   at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.HandleRequestAsync() in Microsoft.AspNetCore.Authentication.dll:token 0x60000c8+0x2e4

   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context) in Microsoft.AspNetCore.Authentication.dll:token 0x6000044+0x1eb

   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task) in Microsoft.AspNetCore.Diagnostics.dll:token 0x60000bf+0x133
```
This weirdly only happens on Firefox on certain clients, only in production (it works in localhost). I've tried disabling all the extensions but I get the same result.
**Steps To reproduce**
Set up Discord as OAuth provider. On Firefox, in production, try to log in to Discord.

**Expected behaviour**
You should log in successfully.

**Actual behaviour**
Getting the following error in console:
```'.AspNetCore.Correlation.wNmPU7i-6bJe81N8aurW-IvEZhT8aJjp1FHuy94h5hY' cookie not found.

Error from RemoteAuthentication: Correlation failed..

An unhandled exception has occurred while executing the request.

System.Exception: An error was encountered while handling the remote login.

 ---> System.Exception: Correlation failed.

   --- End of inner exception stack trace ---

   at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.HandleRequestAsync() in Microsoft.AspNetCore.Authentication.dll:token 0x60000c8+0x2e4

   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context) in Microsoft.AspNetCore.Authentication.dll:token 0x6000044+0x1eb

   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task) in Microsoft.AspNetCore.Diagnostics.dll:token 0x60000bf+0x133
```

**System information:**
 - Hosted on Docker
 - Library Version 5.0.8
 - .NET 6.0 preview 5

**Additional context**
I had the same error using this: https://github.com/Auralytical/Discord.OAuth2

**Comment by martincostello** [comment]
This sounds like it's likely related to SameSite cookies. More information about it can be found here: https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/

The correlation implementation is provided by the base OAuth provider we build on top of, so even if it isn't that, it's unlikely to be an issue specifically with our library, which is further borne out by the fact it isn't working with another library either.

**Comment by martincostello** [comment]
Also, if you're targeting ASP.NET Core 6, I suggest using a prerelease version for 6 from our MyGet feed, rather than the .NET 5 versions in NuGet.

**Comment by NexganGH** [comment]
Thanks for your fast reply @martincostello.

 > This sounds like it's likely related to SameSite cookies. More information about it can be found here: https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/
I had tried to tweak SameSite settings without results.

After some time of testing in localhost, it stopped working there as well (in any browser), getting an `Unknown Location` exception. I found out that when I tried to log in from another page with a redirect it worked but it didn't if I accessed it directly.

So I figured this out: I was using the `account/login` route to log in with an optional parameter which is the return url. It seems like when I try to access directly to that url it actually gets routed to another method which isn't working properly. However, the redirects from the other pages worked and that's because they contained the return url parameter, so Asp.Net routing found my method (having that optional parameter) to be more suitable.

In any case, I just had to rename my route to something else and point to that one instead. Everything is now working (even with Firefox on production), even though I don't know if that was the same problem.
